core/parentchain/indirect-calls-executor/src/indirect_calls_executor.rs

```diff
@@ -143,16 +149,28 @@ where
 						},
 					}
 				}
+				shield_funds_decoded = true;
 			}
+
 			// Found CallWorker extrinsic in block.
-			else if let Ok(xt) = ParentchainUncheckedExtrinsic::<CallWorkerFn>::decode(
-				&mut xt_opaque.encode().as_slice(),
+			// No else-if here! Because the same opaque extrinsic can contain multiple Fns at once (this lead to intermittent M6 failures)
+			if let Ok(xt) = ParentchainUncheckedExtrinsic::<CallWorkerFn>::decode(
```

Comment:
That I still don't understand. The same instance of encoded extrinsic (`encoded_xt_opaque`), can successfully decode into a `ShieldFundsFn` AND a `CallWorkerFn`, suggesting `encoded_xt_opaque` is an aggregate of two unchecked extrinsics. If you know how this is done, please let me know, I would like to write a test for it 😄

Reply:
Hmm, I am actually quite sure that a single extrinsic instance is never an aggregate of two extrinsics. In my understanding, the input data, the encoded value, is not mutated even though we pass a `&mut [...]` reference. So a single instance can decode into both does only mean that it contains valid data for both types it tries to decode into.

Reply:
No, I didn't expect it to mutate anything, that was clear to me. However, your explanation helped me understand one thing now: Just because we can decode successfully, does not mean it also contains valid information.

But then I still don't understand how the same blob of data can successfully decode to `([u8; 2], Vec<u8>, Amount, ShardIdentifier)` AND `([u8; 2], Request)`. They are both tuples, yes, and the first entry is the same. But the rest is different. Or is it that `Request` encoded somehow looks the same as `Vec<u8>, Amount, ShardIdentifier` encoded?

Reply:
Ah I see, I did not even check that the arguments are different, sorry for that.

Yes, the scale-codec is extremely dumb, it contains no contextual information. So being able to decode something does not mean that it contained meaningful data.

`Request` is only a `Vec<u8>` too, so from the args perspective I'd expect:

* `([u8; 2], Vec<u8>, Amount, ShardIdentifier)` can be decoded into both: `CallWorkerFn`, `ShieldFundsFn`
* `([u8; 2], Request)` can be decoded only into `CallWorkerFn`

But from the code I read that  `([u8; 2], Request)` can also decode into both, and this I don't understand why because `Vec<u8>` encoded contains as the first byte the `Vec`s length. So when it tries to decode the `Amount`, there should not be any data left, and return an error.

I had checked the polkadot's wiki, they had a very good explanation for all the types and how they will be encoded, but they are currently migrating it...

Reply:
Since we decode `ShieldFundsFn` first, I think you're right with your assumptions, that would explain the effects I observed. 
We don't see that `([u8; 2], Request)` can be decoded into both. So that would explain what was going on here 😄 thanks for explaining!

Reply:
Ah cool. Happy that we solved the riddle!

Reply:
@[Kailai-Wang](https://github.com/Kailai-Wang): this was the issue behind the failing tests.